File: pull/pull.go
package pull

import (
	"fmt"
	"log"

	"github.com/aura-studio/dynamic-node-cli/config"
)

// Options holds pull configuration.
type Options struct {
	Concurrency int
	Force       bool
	Remote      string // optional: override procedure.warehouse.remote
}

// PullForProcedure downloads warehouse artifacts from remote(s) to local warehouse for a given procedure.
func PullForProcedure(proc config.Procedure, opt Options) {
	if opt.Concurrency <= 0 {
		opt.Concurrency = 8
	}

	remotes := proc.Warehouse.Remote
	if opt.Remote != "" {
		remotes = []string{opt.Remote}
	}

	if len(remotes) == 0 {
		log.Panicf("pull: no warehouse.remote configured")
	}

	name := proc.Target.Namespace + "_" + proc.Target.Package + "_" + proc.Target.Version
	env := proc.Toolchain.OS + "_" + proc.Toolchain.Arch + "_" + proc.Toolchain.Compiler + "_" + proc.Toolchain.Variant

	for _, remote := range remotes {
		r := newRemote(remote)
		count, err := r.PullArtifacts(env, name, proc.Warehouse.Local, opt)
		if err != nil {
			log.Panicf("pull: remote %q failed: %v", remote, err)
		}
		fmt.Printf("Pulled %d file(s) from %s\n", count, remote)
	}
}

File: src/pull/pull.js
import { newRemote } from "./puller.js";

export async function pullForProcedure(proc, opt) {
  if (opt.concurrency <= 0) {
    opt.concurrency = 8;
  }

  let remotes = proc.warehouse.remote;
  if (opt.remote) {
    remotes = [opt.remote];
  }

  if (remotes.length === 0) {
    throw new Error("pull: no warehouse.remote configured");
  }

  const name = `${proc.target.namespace}_${proc.target.package}_${proc.target.version}`;
  const environment = [
    proc.toolchain.os,
    proc.toolchain.arch,
    proc.toolchain.compiler,
    proc.toolchain.variant,
  ].join("_");

  for (const remote of remotes) {
    const source = newRemote(remote);
    const count = await source.pullArtifacts(
      environment,
      name,
      proc.warehouse.local,
      opt,
    );
    console.log(`Pulled ${count} file(s) from ${remote}`);
  }
}
